### PR TITLE
Fetch Key Packages

### LINKED ISSUE #20: Contact Operations – Registry – FetchKeyPackages
### Discussed in https://github.com/xmtp/xps-gateway/discussions/15

<div type='discussions-op-text'>

<sup>Originally posted by **jac18281828** December 22, 2023</sup>
Initial discussion is [here](https://github.com/xmtp/libxmtp/discussions/274)

Device installation is supported by [GrantInstallation](https://github.com/xmtp/xps-gateway/discussions/13). 
Device removal is supported by [RevokeInstallation](https://github.com/xmtp/xps-gateway/discussions/14). 

Retrieval of device current valid installations is supported by `FetchKeyPackages`.

#### Request:

- **Method:** `POST`
- **URL:** `/rpc/v1/fetchKeyPackages`
- **Headers:**
  - `Content-Type: application/json`
- **Body:**
  - **JSON Object:** 
    - `jsonrpc`: `"2.0"`
    - `method`: `"fetchKeyPackages"`
    - `params`: Array (optional parameters as required)
    - `id`: Request identifier (integer or string)

#### Description

The `fetchKeyPackages` endpoint is responsible for retrieving the contact bundle for the XMTP device installations. The request must be made to a valid did with an XMTP profile.

#### Request

The request for this endpoint should contain a valid DID. All returned information is public.

##### Parameters:

-   `DID` (string): Unique XMTP identifier for the user requesting the installation.

##### Example Request:

```json
{
    "jsonrpc": "2.0",
    "method": "fetchKeyPackages",
    "params": {
        "did": "12345"
    },
    "id": 1
}
```

#### Response

The response will provide an optionally empty list of installation bundles.

##### Result Fields:

-   `status` (string): The status of the request, e.g., 'success'.
-   `installation` (array): Array of installation bundles.

##### Example Response:

```json
{
    "jsonrpc": "2.0",
    "result": {
        "status": "success",
        "installation": ["bundle1...", "bundle2..."]
    },
    "id": 1
}
```

#### Error Handling

In case of an error, the response will include an error object with details.

##### Error Object Fields:

-   `code` (integer): Numeric code representing the error type.
-   `message` (string): Description of the error.

##### Example Error Response:

```json
{
    "jsonrpc": "2.0",
    "error": {
        "code": 403,
        "message": "User not authorized for installation."
    },
    "id": 1
}
```

#### Security Considerations

All requests to `fetchKeyPackages` must be made over a secure channel. Ensure that user and application IDs are validated, and proper authentication mechanisms are in place to prevent unauthorized requests.
</div>

### PR DESCRIPTION
**Pull request by insipx**
Closes #20 

requires https://github.com/xmtp/didethresolver/pull/41 to be merged first, for query manipulation

### CODE REVIEW

**Review by jac18281828 — APPROVED** [review]
Thanks this looks really good.  Ready to go, but I would ask what is the definition of 'Completed' in this context.

File: Cargo.toml [hunk 1]
```diff
@@ -22,12 +22,12 @@ serde = "1.0"
 serde_json = "1.0"
 tokio = { version = "1.0", features = ["macros", "rt-multi-thread"] }
 async-trait = "0.1"
-jsonrpsee = { version = "0.21", features = ["macros", "server", "client-core"] }
+jsonrpsee = { version = "0.22", features = ["macros", "server", "client-core"] }
 anyhow = "1.0"
 thiserror = "1.0"
 ethers = { version = "2.0.11", features = ["abigen"] }
 ctor = "0.2"
-lib-didethresolver = { git = "https://github.com/xmtp/didethresolver", package = "lib-didethresolver", branch = "main" }
```
Comment (37ng) [inline Cargo.toml:30]: Curiously if we removed `package = "lib-didethresolver"`, how does cargo locate the package?

File: gateway-types/src/lib.rs [hunk 2]
```diff
@@ -21,23 +23,61 @@ pub struct Message {
...
+    pub status: Status,
+    /// A message relating to the operation
+    pub message: String,
+    /// A list of key packages
+    pub installation: Vec<Bytes>,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
+pub enum Status {
+    Completed,
```
[the remark this answers is not in the document]
Comment (37ng) [reply gateway-types/src/lib.rs]: I second this `Completed` sounds ambiguous to me.

File: registry/src/error.rs [hunk 3]
```diff
@@ -18,4 +18,10 @@ pub enum ContactOperationError<M: Middleware> {
     ProviderError(#[from] ProviderError),
     #[error("Error converting from int: {0}")]
     IntConversion(#[from] TryFromIntError),
+    #[error("Error Resolving {1}: {0}")]
+    ResolutionError(lib_didethresolver::error::ResolverError<M>, String),
+    #[error("The DID has been deactivated, and no longer valid")]
+    DIDDeactivated,
```
Comment (37ng) [inline registry/src/error.rs:24]: Is this **DID**deactivated or it's just certain attribute is deactivated? I think this naming implies that the DID is not longer valid but as long as one has the key, their DID is always valid.

File: xps-gateway/tests/integration_test.rs [hunk 4]
```diff
@@ -266,25 +266,11 @@ async fn test_grant_installation() -> Result<(), Error> {
 #[tokio::test]
 async fn test_revoke_installation() -> Result<(), Error> {
     with_xps_client(None, |client, context, resolver, anvil| async move {
-        let wallet: LocalWallet = anvil.keys()[3].clone().into();
-        let me = get_user(&anvil, 3).await;
+        let me: LocalWallet = anvil.keys()[3].clone().into();
```
Comment (37ng) [inline xps-gateway/tests/integration_test.rs:269]: nice

**Review by 37ng — APPROVED** [review]

File: Cargo.toml (hunk 1 above)
Comment (insipx) [reply Cargo.toml:30]: You can alias packages to a different name, removing the `package` uses the `lib-didethresolver` key as the package name. Cargo uses the virtual manifest located in the directory to locate the package, i.e it checks each directory defined in `[workspace.members]` to see if the Cargo.toml of that package matches

File: registry/src/error.rs (hunk 3 above)
Comment (insipx) [reply registry/src/error.rs:24]: added a test for this

**Pull request merged**